File: src/actions/plan-executor.ts
import { ActionDispatcher } from './dispatcher.js';
import { ConfirmationRepository } from '../database/repositories.js';
import { PlanHasher } from '../security/plan-hasher.js';
import { EntityResolver } from '../entities/entity-resolver.js';
import type { ExecutionContext } from './types.js';
import type { ActionStep } from '../ai/types.js';

export const MAX_PLAN_STEPS = 5;
export const PLAN_TYPE = '__plan__';

export interface PlanStepResult {
  stepId: string;
  action: string;
  ok: boolean;
  result?: any;
  error?: string;
}

export interface PlanConfirmation {
  nonce: string;
  actionType: string;
  riskLevel: string;
  stepLabel: string;
}

export interface PlanExecution {
  /** False when paused for a confirmation; call again after approval. */
  completed: boolean;
  results: PlanStepResult[];
  confirmation?: PlanConfirmation;
}

/** Human-readable plan outcome for embeds. Pure - unit-tested. */
export function summarizePlan(results: PlanStepResult[]): string {
  if (results.length === 0) return 'No steps were executed.';
  const ok = results.filter((r) => r.ok).length;
  const lines = results.map((r) =>
    r.ok ? `✓ ${r.action}` : `✗ ${r.action}: ${(r.error ?? 'failed').slice(0, 120)}`
  );
  return `${ok}/${results.length} steps succeeded.\n${lines.join('\n')}`.slice(0, 1800);
}

/**
 * Executes plan steps in order with the full safety pipeline per step.
 * Name-based params (channelName/roleName) resolve per step like single actions.
 * On a confirmation gate, the REMAINDER (current step onwards) is stashed under
 * a plan confirmation nonce; resume with executeActionPlan(steps, index, ctx).
 * Note: the dispatcher rows its own single-step confirmation first, which then
 * orphans (5-min expiry, harmless) - accepted to avoid duplicating gate policy.
 */
export async function executeActionPlan(
  steps: ActionStep[],
  startIndex: number,
  ctx: ExecutionContext
): Promise<PlanExecution> {
  const results: PlanStepResult[] = [];

  for (let i = startIndex; i < steps.length; i++) {
    const step = steps[i];
    const params: Record<string, any> = { ...(step.parameters ?? {}) };

    if (params.channelName) {
      const res = await EntityResolver.resolveChannel(ctx.guild, params.channelName);
      if (res.resolved) params.channelId = res.resolved.id;
    }
    if (params.roleName) {
      const res = await EntityResolver.resolveRole(ctx.guild, params.roleName);
      if (res.resolved) params.roleId = res.resolved.id;
    }

    const dispatchRes = await ActionDispatcher.dispatch(step.action, params, ctx);

    if (dispatchRes.requiresConfirmation && dispatchRes.confirmationDetails) {
      const plan = { type: PLAN_TYPE, steps, index: i };
      const planHash = PlanHasher.computeHash(plan);
      const actionNonce = PlanHasher.generateNonce();
      await ConfirmationRepository.create({
        guildId: ctx.guildId,
        userId: ctx.userId,
        planHash,
        actionNonce,
        actionPlan: plan,
      });
      return {
        completed: false,
        results,
        confirmation: {
          nonce: actionNonce,
          actionType: step.action,
          riskLevel: dispatchRes.confirmationDetails.riskLevel,
          stepLabel: `${i + 1}/${steps.length}`,
        },
      };
    }

    results.push({
      stepId: step.id,
      action: step.action,
      ok: dispatchRes.success,
      result: dispatchRes.result,
      error: dispatchRes.error,
    });

    if (!dispatchRes.success) {
      return { completed: true, results };
    }
  }

  return { completed: true, results };
}

File: src/actions/plan_executor.py
from __future__ import annotations

from typing import Any

from src.actions.dispatcher import ActionDispatcher
from src.actions.types import ExecutionContext
from src.database.repositories import ConfirmationRepository
from src.entities.entity_resolver import EntityResolver
from src.security.plan_hasher import PlanHasher

MAX_PLAN_STEPS = 5
PLAN_TYPE = "__plan__"


def summarize_plan(results: list[dict[str, Any]]) -> str:
    if not results:
        return "No steps were executed."
    ok = sum(1 for r in results if r.get("ok"))
    lines = [
        f"✓ {r['action']}"
        if r.get("ok")
        else f"✗ {r['action']}: {(r.get('error') or 'failed')[:120]}"
        for r in results
    ]
    return f"{ok}/{len(results)} steps succeeded.\n" + "\n".join(lines)


async def execute_action_plan(
    session: Any,
    steps: list[dict[str, Any]],
    start_index: int,
    ctx: ExecutionContext,
) -> dict[str, Any]:
    """Execute plan steps in order with the full safety pipeline per step.

    Name-based params (channelName/channel_name/roleName/role_name) resolve
    per step. On a confirmation gate the REMAINDER is stashed and a
    confirmation dict is returned; resume with start_index=i after approval.
    """
    results: list[dict[str, Any]] = []

    for i in range(start_index, len(steps)):
        step = steps[i]
        action = step.get("action", "")
        params: dict[str, Any] = dict(step.get("parameters") or {})

        channel_name = params.get("channelName") or params.get("channel_name")
        if channel_name and not params.get("channel_id"):
            res = await EntityResolver.resolve_channel(ctx.guild, str(channel_name))
            if res.resolved is not None:
                params["channel_id"] = str(res.resolved.id)
        role_name = params.get("roleName") or params.get("role_name")
        if role_name and not params.get("role_id"):
            res = await EntityResolver.resolve_role(ctx.guild, str(role_name))
            if res.resolved is not None:
                params["role_id"] = str(res.resolved.id)

        dispatch_res = await ActionDispatcher.dispatch(session, action, params, ctx)

        if dispatch_res.requires_confirmation and dispatch_res.confirmation_details:
            plan = {"type": PLAN_TYPE, "steps": steps, "index": i}
            plan_hash = PlanHasher.compute_hash(plan)
            nonce = PlanHasher.generate_nonce()
            await ConfirmationRepository.create_confirmation(
                session=session,
                guild_id=ctx.guild_id,
                user_id=ctx.user_id,
                plan_hash=plan_hash,
                action_nonce=nonce,
                action_plan=plan,
            )
            return {
                "completed": False,
                "results": results,
                "confirmation": {
                    "nonce": nonce,
                    "action_type": action,
                    "risk_level": dispatch_res.confirmation_details.get("risk_level", "HIGH"),
                    "step_label": f"{i + 1}/{len(steps)}",
                },
            }

        results.append(
            {
                "step_id": step.get("id", str(i)),
                "action": action,
                "ok": dispatch_res.success,
                "result": dispatch_res.result,
                "error": dispatch_res.error,
            }
        )
        if not dispatch_res.success:
            return {"completed": True, "results": results}

    return {"completed": True, "results": results}
